### ISSUE TITLE
Add null safety

### ISSUE THREAD

**Issue by enyo** [mongo-dart/mongo_dart#201]
With the release of dart 2.10 it would be great if this library would add support for this.

With their migration tool, it should be rather easy to get it done.

**Comment by giorgiofran** [comment]
Obviously null safety is an interesting feature, and I believe that we will move in that direction but not before that it will be considered stable (2.10 is still experimental).

**Comment by enyo** [comment]
@giorgiofran the stable channel of dart is now `2.10` and there already is a `2.10.1` release.

**Comment by enyo** [comment]
@giorgiofran Ah i see what you mean. The feature of null safety is apparently still considered experimental.

Would it be an option to already create a `0.4.4-beta` Version?

**Comment by giorgiofran** [comment]
If you read the specs it seems that you can run a null-safety project even if it has dependencies in a non-null-safety one.
It is also stated that it is better to start with the dependencies, because in the "null-safety" conversion, some apis could be changed.
So I did a test starting with the `bson` package, that is a direct dependency of the `mongo_dart` one.
I manually updated the sources and changed all the dependencies that already provided a null-safety pre-release version.
The problem is that `bson` depends on `fixnum` package, that has not yet any null-safety pre-release, and because of that I get errors running the tests.
This is why I'm waiting for a stable "null-safety" functionality. In a recent post, I read that the final version of "null-safety" will be released early next year. This means that we have to wait a little bit yet.

**Comment by enyo** [comment]
That sounds great. Thanks for your feedback. I hope that in a few months, more packages will already have migrated. I'll just leave this issue open so others who are interested in that as well can see it.

**Comment by ab36245** [comment]
I'm really keen to see a null safe version of this excellent package. I see that fixnum now has a null safe preview (1.0.0) available. I'm sure there's significantly more to do than just apply this but is this able to move ahead now?

**Comment by giorgiofran** [comment]
At present I'm working on using the OP_MSG for CRUD operations. Until I reach a stable version I will not start working on null-safety.

**Comment by ab36245** [comment]
That make sense. I'm looking forward to getting the null safe version when it arrives.

**Comment by enyo** [comment]
@giorgiofran this is the last package I need to be able to convert my package to null safety. I appreciate that you're working on something else at the moment, I just wanted to ask whether you had an approximate time frame for when you'll be able to look into this, because if it will take a bit longer, I'll try to find/work on some workaround in the meantime. Thanks!

**Comment by giorgiofran** [comment]
I have not ended the development of the OP_MSG porting, but I did the most and it seems to be quite stable. So I'll stop this thread and start with null-safety as soon as possible.
Please, consider that initially it will be a development version, even because there are still (at least one) dependencies that have not been migrated.

**Comment by paulobreim** [comment]
The problem is that other packages, and new packages already is ade with null_safety.
As we are running flutter stable, null_safety must be considered as stable also.

**Comment by giorgiofran** [comment]
You know, I'm not payed for this. I'm developing this project using my free time (and even more).
This said, I have released a stable version of 0.5.0 (release candidate), so now I can work on null-safety.
I hope to have a development version in 10 days.
This is the best I can do.

**Comment by enyo** [comment]
For what it's worth I donated some money to you a few days ago. It's not much and not going to make any difference (I can't afford more at the moment), I just wanted you to know that your work is much appreciated.

**Comment by paulobreim** [comment]
Very good @giorgiofran. I know that is a hard work.
regards

I can help in tests

paulo

**Comment by giorgiofran** [comment]
If there is any brave developer wishing to test a pre-development release you can do it temporarily using a git reference in your `pubspec.yaml`:
```yaml
dependencies:
  mongo_dart:
    git: 
      url: https://github.com/mongo-dart/mongo_dart.git
      ref: null-safety
```

**Note**: all tests have been passed, but, in some aspects, null-safety is tricky. Maybe some unexpected error can happen.

**Comment by paulobreim** [comment]
> If there is any brave developer wishing to test a pre-development release you can do it temporarily using a git reference in your `pubspec.yaml`:
> 
> ```yaml
> dependencies:
>   mongo_dart:
>     git: 
>       url: https://github.com/mongo-dart/mongo_dart.git
>       ref: null-safety
> ```
> 
> **Note**: all tests have been passed, but, in some aspects, null-safety is tricky. Maybe some unexpected error can happen.

I Tested with 2 Db and 3 DbCollection.
Find and Insert functions.

I read a mongo with 3 millions registers using the code bellow.

 sql = '0000000000-0';
 result =  await coll_Historia.find(where.gt('_id', sql).limit(1)).toList();
 while (result.isNotEmpty) {
    sql = result[0]['_id'];  // get sql to next find gt
    // do something
  result =  await coll_Historia.find(where.gt('_id', sql).limit(1)).toList();
}

Taking the opportunity...
Are there an away to read register by register in different form?
look that I have 3 millions of register



}

**Comment by enyo** [comment]
Thank you so much. That is worth a lot to me because I can continue working on my server now.

**Comment by giorgiofran** [comment]
@paulobreim may you please open a separated issue, so it will remain a trace also for other users?

**Comment by enyo** [comment]
@giorgiofran I've used the new library now, and all my tests pass without an issue, so I haven't noticed any problem with it so far. If I discover something I'll report it.

**Comment by ab36245** [comment]
I have also run my set of tests and it all checks out so far. Like @enyo I'll keep checking and report anything I find. Thanks very much @giorgiofran

**Comment by giorgiofran** [comment]
Thanks guys! This is a good start. I have found a bug, but it was already there, it didn't depend on null safety.
I have updated the repository, please update your dependencies.
I'd like to make some more tests, I hope to have time this week-end. 
After that I will publish a dev release.

**Comment by ab36245** [comment]
Works for my tests. My usage is fairly simple, so I may not be testing the bug, but at least no regressions! :-)

**Comment by enyo** [comment]
@giorgiofran the only thing that seems to have stopped working, is that I can't connect to my GitHub action mongodb service anymore.

```
  MongoDB ConnectionException: Could not connect to mongodb:27017
  - SocketException: Failed host lookup: 'mongodb' (OS Error: Temporary failure in name resolution, errno = -3)
  package:mongo_dart/src/network/connection_manager.dart 94:9  _ConnectionManager.open
```

I didn't have that problem before I upgraded, but I have made a few other changes as well, so it might be unrelated to `mongo_dart`, but I have spent some time testing and debugging and I can't seem to find another reason. Did you change something to the networking and is it possible that a change in the library is the cause of this issue?
